### ISSUE TITLE
RegisterConditional with custom InjectionConsumerInfo

### ISSUE THREAD

**Issue by DaEkstrim** [simpleinjector/SimpleInjector#237]
We have a multi tenancy application which is using Dependency Injection to provide services depending on some attribute information the implementation is decorated with.

I was taking a look at SimpleInjector and I found that it supports RegisterConditional, where one of the overloads accepts Func<TypeFactoryContext, Type>. While taking a look at TypeFactoryContext, I came across the Consumer which is of type InjectionConsumerInfo which is a public unsealed class.

My question is: Is it possible to provide the conditional predicate with our custom class that inherits from InjectionConsumerInfo that provides a little more information (related to the multi tenancy in question such as ClientId)? If yes, how can this be achieved?

**Comment by dotnetjunkie** [comment]
Can you supply us with concrete examples of what you are trying to accomplish?

**Comment by DaEkstrim** [comment]
```
[InjectableFor(ClientId: 1)]
public class Client1Service: ISerivce{}

[InjectableFor(ClientId: 2)]
public class Client2Service: IService{}

container.RegisterConditional(typeof(IService), typeof(Client1Service), c => (c.Consumer as OurCustomInjectionConsumerInfo).ClientId == 1);
```

Please treat this code as pseudocode. In reality we would loop through all available implementations that have the InjectableFor attribute and create a RegisterConditional for each implementation. What I'm not sure is whether we can provide the predicate with our own `InjectionConsumerInfo` which would contain the necessary information to execute the condition.

**Comment by dotnetjunkie** [comment]
So how do you know when you're running in the context of Client1 or not?

**Comment by DaEkstrim** [comment]
This approach is going to be used in a WebAPI app which serves requests on multiple http address bindings. Depending on which address the request was received, we will know which client code needs to run. So if the request is received on www.client1.com, we know that the ClientId is 1, we build the context object and then provide it to the container for the duration of the request which will in turn serve implementations whose ClientId matches the one in the context object.

If I'm not being clear, please do tell me. I'm doing my best to provide all need to know information without overwhelming the post with irrelevant details.

**Comment by dotnetjunkie** [comment]
The information is clear. I've got all ingormation I need, but I need some time to formulate a good response. I'll try answer your question later this weekend.

**Comment by Tornhoof** [comment]
This does not directly solve your problem, but might help you finding your solution. I once wrote for a multi-tenant application a per-tenant singleton lifestyle (for per tenant caching): https://gist.github.com/Tornhoof/9014ed743f1edce0abe4afb49f671783 (the actual per tenant stuff is the concurrentdictionary with identityhelper)
This does not resolve different implementation types, but only one, but the parts where it returns the implementation type could probably be adopted based on your registration attributes.

**Comment by kaagati** [comment]
I don't think this is possible using SimpleInjector. The problem is, Instance should be materialized before it is available to use. The best I could have tried would be create a Factory that would have a method GetService that returns IService based on the condition and inject that instead of IService.

**Comment by dotnetjunkie** [comment]
Compared to what kaagati suggests, there's no technical limitation in Simple Injector that prevents you from doing what you want, and I could describe how to create the proper registrations, but I that would not solve some deeper issues will still be having. So instead I would like to feedback on what you want to do.

What you want to do is let the structure of your object graph depend on data that is only available at runtime (the http address in your case). Just as [it is a bad idea to inject runtime data into application components](https://www.cuttingedge.it/blogs/steven/pivot/entry.php?id=99), it is a bad idea to construct object graphs based on runtime data, for the same reasons: it complicates the object graph and makes it impossible to verify you composition root. Instead, your object graph should be fixed in structure and the use of runtime data should be delayed till after the object graph is built, and used to select the method call graph through the existing object graph.

This means that in your case you will typically need some sort of ITenantContext service that allows returning the information of the currently executing Tenant. For Web API such implementation might look as follows:

``` c#
private sealed class SimpleInjectorWebApiTenantContext : ITenantContext {
    private readonly Container container;

    public SimpleInjectorWebApiTenantContext(Container container) {
        this.container = container;
    }

    public string TenantName => GetTeantNameFromUri(CurrentMessage.RequestUri);

    private HttpRequestMessage CurrentMessage => container.GetCurrentHttpRequestMessage();
}
```

And you can register this class as follows:

``` c#
var container = new Container();
container.EnableHttpRequestMessageTracking(GlobalConfiguration.Configuration);

container.RegisterSingleton<ITenantContext>(new SimpleInjectorWebApiTenantContext(container));
```

This leaves you with the proper `IService` implementation based on the current tenant. There are multiple solutions here, what to choose depends on the amount of flexibility you require, but a simple solution is to implement a proxy that forwards call to injected implementations:

``` c#
public class TenantServiceProxy : ISerivce
{
    private readonly ITenantContext tenantContext;
    private readonly Client1Service client1Service;
    private readonly Client1Service client2Service;
    private readonly DefaultService defaultService;

    public TenantServiceProxy(ITenantContext tenantContext, 
        Client1Service client1Service,
        Client1Service client2Service,
        DefaultService defaultService) {
        this.tenantContext = tenantContext;
        this.client1Service = client1Service;
        this.client2Service = client2Service;
        this.defaultService = defaultService;
    }

    public object ServiceMethod(object arg) {
        return GetTenantService().ServiceMethod(arg);
    }

    private IService GetTenantService() {
        switch (this.tenantContext.TenantName) {
            case "Client1": return client1Service;
            case "Client2": return client2Service;
            default: return defaultService;
        }
    }

}
```

This builds the complete object graph up front and allows every service to be validated verified and analyzed. The only thing you need to do is registering it as follows:

``` c#
container.Register<IService, TenantServiceProxy>();
```

The `TenantServiceProxy` can be part of your composition root where you have all information available about which tenants you run. Downside of this approach is of course that you have constant maintenance on this `TenantServiceProxy` in case you are constantly adding new tenant-specific `IService` implementations. In that case, you're better of again with a model where your proxy is configured with a dictionary containing `InstanceProducer` instances (or something similar), that allows you to do a callback to the container at runtime to get that part of the object graph. For instance:

``` c#
public class TenantServiceProxy : ISerivce
{
    private readonly ITenantContext tenantContext;
    private readonly Dictionary<string, InstanceProducer> services;

    public TenantServiceProxy(ITenantContext tenantContext, 
        Dictionary<string, InstanceProducer<> services) {
        this.tenantContext = tenantContext;
        this.services = services;
    }

    public object ServiceMethod(object arg) {
        return GetTenantService().ServiceMethod(arg);
    }

    private IService GetTenantService() =>
        services.ContainsKey(this.tenantContext.TenantName)
            ? (IService)services[this.tenantContext.TenantName].GetInstance()
            : (IService)services["Default"].GetInstance();
    }
}
```

Building up the list of `InstanceProducers` can be done using reflection as you are already doing and by calling `Lifestyle.Transient.CreateProducer(typeof(IService), typeof(Client1Service), container)` for instance. You can register everything as follows:


``` c#
var container = new Container();
container.EnableHttpRequestMessageTracking(GlobalConfiguration.Configuration);

ITenantContext context = new SimpleInjectorWebApiTenantContext(container);

container.RegisterSingleton<ITenantContext>(context);

Dictionary<string, InstanceProducer> services = CreateServices(container);

container.RegisterSingleton<IService>(new TenantServiceProxy(context, services));
```

**Comment by DaEkstrim** [comment]
Thanks a lot for the feedback @dotnetjunkie. I'm in awe at the depth and quality of your response. I have run your post with the team and your suggested proxy approach to the problem, and we all agreed that this solution addresses our core concerns. We might even go one step further and use interception with dynamically typed proxies that are use the `ITenantContext` to deduce to which implementation the method calls will be forwarded to.

Which in the end means that even though this was not really a SimpleInjector issue but rather a pattern of approach to the problem, you still took your time to provide us with a solution. And for that we are truly grateful. Hopefully future readers will find this post helpful as well. Thanks once again.
